File: daily_problems/2026/06/0605/personal_submission/cf102860b_yawn_sean.py
# Submission link: https://codeforces.com/gym/102860/submission/377278085
def main():
    n, l = MII()
    nums = LII()
    nums.sort()
    
    for i in range(n):
        nums.append(nums[i] + l)
    
    ans = n * (n - 1) * (n - 2) // 6
    r = 0
    
    for i in range(n):
        while 2 * (nums[r] - nums[i]) < l:
            r += 1
        
        v = r - i - 1
        ans -= v * (v - 1) // 2
    
    print(ans)

File: daily_problems/2026/06/0605/personal_submission/cf102860b_yawn_sean.cpp
#include <bits/stdc++.h>
// #pragma GCC optimize("O3,Ofast,unroll-loops")
// #include "atcoder/all"

using namespace std;
auto rng = mt19937(random_device()());
auto rngl = mt19937_64(random_device()());

int main() {
	ios_base::sync_with_stdio(false);
	cin.tie(0);
	cout.tie(0);

	int n, l;
	cin >> n >> l;

	vector<int> nums(n);
	for (auto &x: nums) cin >> x;

	sort(nums.begin(), nums.end());
	for (int i = 0; i < n; i ++) nums.emplace_back(nums[i] + l);

	int r = 0;
	long long ans = 1ll * n * (n - 1) * (n - 2) / 6;

	for (int i = 0; i < n; i ++) {
		while (2 * (nums[r] - nums[i]) < l) r ++;
		int v = r - i - 1;
		ans -= 1ll * v * (v - 1) / 2;
	}

	cout << ans;

	return 0;
}